app/derivative_services/vips_derivative_service.rb

```diff
@@ -0,0 +1,167 @@
+# frozen_string_literal: true
+class VIPSDerivativeService
+  # Pixel width or height at which point it cuts the size in half for
+  # performance.
+  REDUCTION_THRESHOLD = 15_000
+  class Factory
+    attr_reader :change_set_persister
+    delegate :metadata_adapter, to: :change_set_persister
+    delegate :query_service, to: :metadata_adapter
+    def initialize(change_set_persister:)
+      @change_set_persister = change_set_persister
+    end
+
+    def new(id:)
+      VIPSDerivativeService.new(id: id, change_set_persister: change_set_persister)
+    end
+  end
+
+  class IoDecorator < SimpleDelegator
+    attr_reader :original_filename, :content_type, :use
+    def initialize(io, original_filename, content_type, use)
+      @original_filename = original_filename
+      @content_type = content_type
+      @use = use
+      super(io)
+    end
+  end
+
+  attr_reader :change_set_persister, :id
+  delegate :mime_type, to: :target_file
+  delegate :query_service, to: :change_set_persister
+  def initialize(id:, change_set_persister:)
+    @id = id
+    @change_set_persister = change_set_persister
+  end
+
+  def resource
+    @resource ||= query_service.find_by(id: id)
+  end
+
+  def target_file
+    @target_file ||= intermediate_target_files(resource) || resource.original_file
+  end
+
+  # If there are intermediate files with the supported format attached to the
+  #   resource, select the first of these
+  # @param [Valkyrie::Resource] resource
+  # @return [FileMetadata]
+  def intermediate_target_files(resource)
+    supported = resource.intermediate_files.select do |intermed|
+      ["image/tiff", "image/jpeg"].include?(intermed.mime_type.first)
+    end
+    supported.empty? ? nil : supported.first
+  end
+
+  def valid?
+    ["image/tiff", "image/jpeg"].include?(mime_type.first)
+  end
+
+  def create_derivatives
+    run_derivatives
+    change_set.files = [build_file]
+    change_set_persister.buffer_into_index do |buffered_persister|
+      @resource = buffered_persister.save(change_set: change_set)
+    end
+    update_error_message(message: nil) if target_file.error_message.present?
+  rescue StandardError => error
+    update_error_message(message: error.message)
+    raise error
+  end
+
+  def change_set
+    @change_set ||= DynamicChangeSet.new(resource)
+  end
+
+  def run_derivatives
+    vips_image.tiffsave(
+      temporary_output.path.to_s,
+      compression: :jpeg,
+      tile: true,
+      pyramid: true,
+      Q: 75,
+      tile_width: 1024,
+      tile_height: 1024,
+      strip: true,
+      profile: color_profile
+    )
+    raise "Unable to store pyramidal TIFF for #{filename}!" unless File.exist?(temporary_output.path)
+  end
+
+  def vips_image
+    @vips_image ||=
+      begin
+        image = Vips::Image.new_from_file(filename.to_s)
+        if image.height >= REDUCTION_THRESHOLD || image.width >= REDUCTION_THRESHOLD
+          image.resize(0.5)
```

Comment:
I wonder if it would be better to set the scale factor proportionally based on the pixels above the REDUCTION_THRESHOLD. For example, if there is an image with a height of 31,000 pixels, the resized image will still be above 15,000. Such large images are probably not that common, I would image, so maybe it's not worth the trouble.